ucp/core.py

```diff
@@ -472,28 +519,23 @@ class Endpoint:
     to create an Endpoint.
     """
 
-    def __init__(
-        self,
-        endpoint,
-        ctx,
-        msg_tag_send,
-        msg_tag_recv,
-        ctrl_tag_send,
-        ctrl_tag_recv,
-        guarantee_msg_order,
-    ):
+    def __init__(self, endpoint, ctx, guarantee_msg_order, tags=None):
         self._ep = endpoint
         self._ctx = ctx
-        self._msg_tag_send = msg_tag_send
-        self._msg_tag_recv = msg_tag_recv
-        self._ctrl_tag_send = ctrl_tag_send
-        self._ctrl_tag_recv = ctrl_tag_recv
         self._guarantee_msg_order = guarantee_msg_order
         self._send_count = 0  # Number of calls to self.send()
         self._recv_count = 0  # Number of calls to self.recv()
         self._finished_recv_count = 0  # Number of returned (finished) self.recv() calls
         self._shutting_down_peer = False  # Told peer to shutdown
         self._close_after_n_recv = None
+        if tags is not None:
+            self._use_tags = True
+            self._msg_tag_send = tags["msg_tag_send"]
+            self._msg_tag_recv = tags["msg_tag_recv"]
+            self._ctrl_tag_send = tags["ctrl_tag_send"]
+            self._ctrl_tag_recv = tags["ctrl_tag_recv"]
+        else:
+            self._use_tags = False
```

Comment:
What about keeping the dict of tags like:
```suggestion
        self._tags = tags        
```
Than `self._tags==None` means _no tags_

Reply:
I did think about that, the main though is that it would require an additional dict lookup for every tag operation, which my MPI thinking brain says, "4 byte sends are valid, and some people care a lot about that kind of injection rate"

Reply:
I think we can afford dict lookup in this case. Alternatively, we could create a `struct` of the tag values that is `None` when tags aren't used.